src/EventStore.Core/Services/Storage/EpochManager/EpochManager.cs

```diff
@@ -177,101 +211,139 @@ public class EpochManager : IEpochManager {
-				                                  + "SetLastEpoch FAILED! Data corruption risk!",
-					epoch.EpochPosition,
-					epoch.EpochNumber,
-					epoch.EpochId));
+				throw new Exception(
+					$"Not found epoch at {epoch.EpochPosition} with epoch number: {epoch.EpochNumber} and epoch ID: {epoch.EpochId}. " +
+					"SetLastEpoch FAILED! Data corruption risk!");
 			}
 		}
+		public void InternalCacheEpoch(EpochRecord epoch) {
+			Ensure.NotNull(epoch, "epoch");
 
-		private void UpdateLastEpoch(EpochRecord epoch, bool flushWriter) {
 			lock (_locker) {
-				_epochs[epoch.EpochNumber] = epoch;
-				_lastEpochNumber = epoch.EpochNumber;
-				_lastEpochPosition = epoch.EpochPosition;
-				_minCachedEpochNumber = Math.Max(_minCachedEpochNumber, epoch.EpochNumber - CachedEpochCount + 1);
-				_epochs.Remove(_minCachedEpochNumber - 1);
-
-				if (flushWriter)
-					_writer.Flush();
-				// Now update epoch checkpoint, so on restart we don't scan sequentially TF.
-				_checkpoint.Write(epoch.EpochPosition);
-				_checkpoint.Flush();
-
-				Log.Debug(
-					"=== Update Last Epoch E{epochNumber}@{epochPosition}:{epochId:B} (previous epoch at {lastEpochPosition}) L={leaderId:B}.",
+				
+				// if we have a first or last and either are equal it's already cached, just return
+				if ((_epochs.Last?.Value.EpochNumber ?? -1) == epoch.EpochNumber ||
+					(_epochs.First?.Value.EpochNumber ?? -1) == epoch.EpochNumber) { return; } 
+
+				//new last epoch written or received, this is the normal case
+				//if the list is empty Last will be null
+				if (_epochs.Last == null || _epochs.Last.Value.EpochNumber < epoch.EpochNumber) {
+					_epochs.AddLast(epoch);
+					_lastCachedEpoch = _epochs.Last;
+					_firstCachedEpoch = _epochs.First;
+					// Now update epoch checkpoint, so on restart we don't scan sequentially TF.
+					_checkpoint.Write(_epochs.Last.Value.EpochPosition);
+					_checkpoint.Flush();
+					Log.Debug(
+						"=== Cached new Last Epoch E{epochNumber}@{epochPosition}:{epochId:B} (previous epoch at {lastEpochPosition}) L={leaderId:B}.",
+						epoch.EpochNumber, epoch.EpochPosition, epoch.EpochId, epoch.PrevEpochPosition, epoch.LeaderInstanceId);
+					return;
+				}
+
+				//see if we can avoid iterating list and just prepend 
+				if (epoch.EpochNumber < _epochs.First.Value.EpochNumber) {
+					if (_epochs.Count >= _cacheSize) { return; }//but only if space is available
+					//add new first epoch
+					_epochs.AddFirst(epoch);
+					_firstCachedEpoch = _epochs.First;
+					Log.Debug(
+						"=== Cached Epoch E{epochNumber}@{epochPosition}:{epochId:B} (previous epoch at {lastEpochPosition}) L={leaderId:B}.",
+						epoch.EpochNumber, epoch.EpochPosition, epoch.EpochId, epoch.PrevEpochPosition, epoch.LeaderInstanceId);
+					return;
+				}
+
+				//List insert edge case, this is likely a race condition near the last epoch
+				var e = _epochs.Last;
+				while (e?.Previous != null) {
+					if (e.Previous.Value.EpochNumber < epoch.EpochNumber) {
```

Comment:
@condron duplicates are not being verified here -  the same epoch record can be added twice to the list

Reply:
addressed